source/jobs/JobsFeature.cpp

```diff
@@ -534,121 +528,137 @@ bool JobsFeature::isDuplicateNotification(JobExecutionData job)
+        const int actualPermissions = FileUtils::GetFilePermissions(commandStream.str().c_str());
         if (Permissions::JOB_HANDLER != actualPermissions)
         {
             string message = FormatMessage(
                 "Unacceptable permissions found for job handler %s, permissions should be %d but found %d",
-                Sanitize(command.str()).c_str(),
+                Sanitize(commandStream.str()).c_str(),
                 Permissions::JOB_HANDLER,
                 actualPermissions);
             LOG_ERROR(TAG, message.c_str());
-            publishUpdateJobExecutionStatus(job, {Iotjobs::JobStatus::FAILED, message, "", ""});
-            return;
+            throw std::runtime_error(message);
         }
     }
 
-    int allowStdErr = 0;
-    if (jobDoc.KeyExists(JOB_ATTR_ALLOW_STDERR))
-    {
-        allowStdErr = jobDoc.GetInteger(JOB_ATTR_ALLOW_STDERR);
-    }
+    return commandStream.str();
+}
 
-    LOGM_DEBUG(
-        TAG, "About to execute: %s %s", Sanitize(command.str()).c_str(), Sanitize(argsStringForLogging.str()).c_str());
-    unique_ptr<JobEngine> engine(new JobEngine);
-    int executionStatus = engine->exec_cmd(command.str().c_str(), args);
+void JobsFeature::executeJob(JobExecutionData job)
+{
+    LOGM_INFO(TAG, "Executing job: %s", job.JobId->c_str());
+    publishUpdateJobExecutionStatus(job, {Iotjobs::JobStatus::IN_PROGRESS, "", "", ""});
 
-    ostringstream reason;
-    if (WIFEXITED(executionStatus))
-    {
-        reason << "Exited with status: " << WEXITSTATUS(executionStatus);
-    }
-    else if (WIFSIGNALED(executionStatus))
+    Aws::Crt::JsonView jobDoc = job.JobDocument->View();
+    Aws::Crt::String operation = jobDoc.GetString(JOB_ATTR_OP);
+    Aws::Crt::String path = jobDoc.GetString(JOB_ATTR_PATH);
+    bool includeStdOut = jobDoc.KeyExists(JOB_ATTR_INCLUDE_STDOUT) ? jobDoc.GetBool(JOB_ATTR_INCLUDE_STDOUT) : false;
+
+    vector<string> args;
+    ostringstream argsStringForLogging;
+    if (jobDoc.KeyExists(JOB_ATTR_ARGS) && jobDoc.GetJsonObject(JOB_ATTR_ARGS).IsListType())
     {
-        reason << "Killed by signal: " << WTERMSIG(executionStatus);
+        for (Aws::Crt::JsonView view : jobDoc.GetArray(JOB_ATTR_ARGS))
+        {
+            args.push_back(view.AsString().c_str());
+            argsStringForLogging << view.AsString().c_str() << " ";
+        }
     }
-    else if (WIFSTOPPED(executionStatus))
+    else
     {
-        reason << "stopped by signal: " << WSTOPSIG(executionStatus);
+        LOG_INFO(
+            TAG, "Did not find any arguments in the incoming job document. Value should be a JSON array of arguments");
     }
-    else
+
+    auto shutdownHandler = [this]() -> void {
+        handlingJob.store(false);
+        if (needStop.load())
+        {
+            LOGM_INFO(TAG, "Shutting down %s now that job execution is complete", getName().c_str());
+            baseNotifier->onEvent((Feature *)this, ClientBaseEventNotification::FEATURE_STOPPED);
```

Comment:
I think this might lead to a situation where the baseNotifier fires the the ClientBaseEventNotification::FEATURE_STOPPED if JobsFeature::stop where to execute in parallel with this method, although I'm not sure if thread model allows parallel execution of those two methods or not.

Reply:
I think it may technically be possible, but there are some thread safety mechanisms on the other side within the event handler (`handle_feature_stopped`) that increase thread safety. I can't currently identify any bad consequences if the event notifier were to fire twice, other than some wasted CPU cycles